aws/src/main/java/org/apache/iceberg/aws/s3/VendedCredentialsProvider.java

```diff
@@ -0,0 +1,141 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws.s3;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.ErrorHandlers;
+import org.apache.iceberg.rest.HTTPClient;
+import org.apache.iceberg.rest.RESTClient;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+import org.apache.iceberg.rest.auth.OAuth2Util;
+import org.apache.iceberg.rest.credentials.Credential;
+import org.apache.iceberg.rest.responses.LoadCredentialsResponse;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+public class VendedCredentialsProvider implements AwsCredentialsProvider, SdkAutoCloseable {
+  public static final String URI = "credentials.uri";
+  private volatile HTTPClient client;
+  private final Map<String, String> properties;
+  private final CachedSupplier<AwsCredentials> credentialCache;
+
+  private VendedCredentialsProvider(Map<String, String> properties) {
+    Preconditions.checkArgument(null != properties, "Invalid properties: null");
+    Preconditions.checkArgument(null != properties.get(URI), "Invalid URI: null");
+    this.properties = properties;
+    this.credentialCache =
+        CachedSupplier.builder(this::refreshCredential)
+            .cachedValueName(VendedCredentialsProvider.class.getName())
+            .build();
+  }
+
+  @Override
+  public AwsCredentials resolveCredentials() {
+    return credentialCache.get();
+  }
+
+  @Override
+  public void close() {
+    IoUtils.closeQuietly(client, null);
+    credentialCache.close();
+  }
+
+  public static VendedCredentialsProvider create(Map<String, String> properties) {
+    return new VendedCredentialsProvider(properties);
+  }
+
+  private RESTClient httpClient() {
+    if (null == client) {
+      synchronized (this) {
+        if (null == client) {
+          client = HTTPClient.builder(properties).uri(properties.get(URI)).build();
+        }
+      }
+    }
+
+    return client;
+  }
+
+  private LoadCredentialsResponse fetchCredentials() {
+    return httpClient()
+        .get(
+            properties.get(URI),
+            null,
+            LoadCredentialsResponse.class,
+            OAuth2Util.authHeaders(properties.get(OAuth2Properties.TOKEN)),
```

Comment:
Assuming the Token here is same short lived token used by RestCatalog instance, are we planning to handle token refresh post its expiration with in the FileIO?

Reply:
this is indeed an issue that I was planning to address when I wrote the first version of this many months ago but simply forgot to get back to. thanks for raising this @ChaladiMohanVamsi

Reply:
I think for the initial support for refresh, we can assume that we're bound by the token lifetime for refreshes.  It's not perfect, but I'd like to see the resolution of some of the AuthManger refactor before settling on a solution.  We're not regressing at this point, so I'm ok with leaving this as is and addressing in the future.

Reply:
FYI, this is on my todo list for after the AuthManager refactor.